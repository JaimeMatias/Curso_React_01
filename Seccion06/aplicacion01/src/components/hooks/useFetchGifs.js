import { useEffect, useState } from "react"
import { getGifs } from "../../helpers/getGifs"

export const useFetchGifs = (category) => {
    console.log('Valor en UseFetchGifs',category)
    const [state, setState] = useState({
        data: [],
        loading: true
    })


    useEffect(() => {
        console.log('El valor a enviar a GetGif',category)
        getGifs(category).then(imgs => {
            setTimeout(() => {
                setState({
                    data: imgs,
                    loading: false
                })

            }, 3000)


        }
        )
    }, [category]
    )

    return state
}