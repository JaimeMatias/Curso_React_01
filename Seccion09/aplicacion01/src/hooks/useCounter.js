import { useState } from "react"

export const useCounter = (initialState=10) => {
  const[state,setState]=useState(initialState);

  const increment =()=>{
      setState(state+1);
  };

  const decrement =()=>{
    setState(state-1);
}
const reset=()=>{
    setState(initialState);
}
    return {state,increment,decrement,reset}
    //el Return devuelve 4 objetos
    //State es el hook
    //increment, es una referencia a la función que aumenta el valor del Hook y en donde se importe se la llama simplemente con ese nombre 'increment'
    //decrement, es una referencia a la función que decrementa el valor del Hook y en donde se importe se la llama simplemente con ese nombre 'decrement'
    //reset, es una referencia a la función que reinicia el valor del Hook y en donde se importe se la llama simplemente con ese nombre 'reset'
}



