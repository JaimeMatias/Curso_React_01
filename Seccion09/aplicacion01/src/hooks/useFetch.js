import { useEffect, useState } from "react"

export const useFech = (url) => {
    const [state, setState] = useState({ data: null, loading: true, error: true });//El primer valor del state se define ac치
    useEffect(() => {
        setState({data:null,loading:true,error:null}) //El valor de state mientras carga se muesta ac치
        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                setState({
                    loading: false,
                    error: null,
                    data
                })//Una vez cargado el componente, el state se actualiz치 ac치
            })
    }, [url])
    return state
}