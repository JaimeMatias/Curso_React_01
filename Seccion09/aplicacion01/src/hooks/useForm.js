import React, { useState } from 'react'

const useForm = (initialState={}) => {
  const [values, setValues]=useState(initialState)

  const handleInputChange =(e)=>{
    //console.log(e.target)
    console.log(e.target.name)
    console.log(e.target.value)
    setValues({
        ...values,
        [e.target.name]: e.target.value
    })
  }

  return[values,handleInputChange]
}

export default useForm
