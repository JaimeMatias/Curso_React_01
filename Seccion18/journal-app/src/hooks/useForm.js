import { useEffect, useState } from 'react';

export const useForm = (initialForm = {}, formValidations = {}) => {

    const [formState, setFormState] = useState(initialForm);
    const [formValidation, setFormValidation] = useState({})
    useEffect(() => {

        createValidators();
        console.log(formValidation)
    }, [formState])

    const onInputChange = ({ target }) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [name]: value
        });
    }

    const onResetForm = () => {
        setFormState(initialForm);
    }
    const createValidators = () => {
        const formCheckedValues = {};
        console.log(Object.keys(formValidations))
        for (const formField of Object.keys(formValidations)) {

            const [fn, errorMessage] = formValidations[formField]; // Extrae los valores de cada elemento, pero les pone un nombre tambien, lo llama fn y errorMessage
            console.log(`${formState[formField]} ${fn(formState[formField])}`)
            formCheckedValues[`${formField}Valid`] = fn(formState[formField]) ? null : errorMessage
        }
        setFormValidation(formCheckedValues)
     
    }
    return {
        ...formState,
        formState,
        onInputChange,
        onResetForm,
        ...formValidation,
    }
}