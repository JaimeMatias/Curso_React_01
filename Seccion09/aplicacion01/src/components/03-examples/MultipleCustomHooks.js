import React from 'react'
import { useFech } from '../../hooks/useFetch'

const MultipleCustomHooks = () => {
    const url='https://www.breakingbadapi.com/api/quotes/1' //
    
    const state=useFech(url)
console.log(state)
  return (
    <div>
      <h1>Custom Hooks!!!</h1>
    </div>
  )
}

export default MultipleCustomHooks
