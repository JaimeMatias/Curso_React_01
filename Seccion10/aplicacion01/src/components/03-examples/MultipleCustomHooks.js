import React from 'react'
import { useFech } from '../../hooks/useFetch'
import { useCounter } from "../../hooks/useCounter";




let vaible=0
const MultipleCustomHooks = () => {
  //Un componente puede recargase varias veces sin que el usuario realice ninguna modificaci√≥n, este componente se ejecuta 3 veces
  const defaulte = 1;
  
  const { state, increment, decrement, reset } = useCounter(defaulte)
  const url = `https://www.breakingbadapi.com/api/quotes/${state}`

  const { loading, data } = useFech(url)
  vaible++

  const { author, quote } = !!data && data[0];

  return (
    <div>
      <h1>BreakingBad Quotes</h1>
      <hr />
      {
        loading
          ? (
            <div className='alert alert-info text-certer'>
              Loading...
            </div>
          )
          : (
            <blockquote className='blockquote text-right'>
              <p className='mb-0' >{author}</p>
              <footer className='blockquote-foot'>{quote}</footer>
            </blockquote>

          )
      }
      <button onClick={decrement} className="btn">anterior</button>
      <button onClick={reset} className="btn">reiniciar</button>
      <button onClick={increment} className="btn">siguiente</button>



    </div>
  )
}

export default MultipleCustomHooks
