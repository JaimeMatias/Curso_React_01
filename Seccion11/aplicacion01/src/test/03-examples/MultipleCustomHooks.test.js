import React from 'react';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../components/03-examples/MultipleCustomHooks';
import { useFech } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';
jest.mock('../../hooks/useFetch') //Genera un mock del useFetch
jest.mock('../../hooks/useCounter') //Genera un mock del useCounter


describe('Pruebas en <MultipleCustomHooks />', () => {
    useCounter.mockReturnValue({
        state: 10,
        increment: () => { }
    })

    test('debe de mostrarse correctamente ', () => {
        useFech.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })
        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    })

    test('Debe de mostrar la informaciÃ³n', () => {

        useFech.mockReturnValue({
            data: [{
                author: 'Fernando',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper.find('.alert').exists()).toBe(false)
        expect(wrapper.find('footer').text().trim()).toBe('Hola Mundo')
        expect(wrapper.find('.mb-0').text().trim()).toBe('Fernando')
    })


})
