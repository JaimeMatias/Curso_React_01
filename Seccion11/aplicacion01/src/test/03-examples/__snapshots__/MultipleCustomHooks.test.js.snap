// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe de mostrarse correctamente  1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-certer"
  >
    Loading...
  </div>
  <button
    className="btn"
  >
    anterior
  </button>
  <button
    className="btn"
  >
    reiniciar
  </button>
  <button
    className="btn"
    onClick={[Function]}
  >
    siguiente
  </button>
</div>
`;

exports[`Pruebas en <MultipleCustomHooks />> debe de mostrarse correctamente  1`] = `
<div>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-certer"
  >
    Loading...
  </div>
  <button
    className="btn"
    onClick={[Function]}
  >
    anterior
  </button>
  <button
    className="btn"
    onClick={[Function]}
  >
    reiniciar
  </button>
  <button
    className="btn"
    onClick={[Function]}
  >
    siguiente
  </button>
</div>
`;
